packages/mcfunction/src/parser/macro.ts

```diff
@@ -0,0 +1,161 @@
+import * as core from '@spyglassmc/core'
+import { localeQuote, localize } from '@spyglassmc/locales'
+import type { MacroChildNode, MacroKeyNode, MacroNode } from '../node/index.js'
+
+/**
+ * Parse a macro line.
+ */
+export function macro(): core.Parser<MacroNode> {
+	return (
+		src: core.Source,
+		ctx: core.ParserContext,
+	): core.Result<MacroNode> => {
+		const start = src.cursor
+		const children: MacroChildNode[] = []
+
+		// Handle the starting '$'
+		const sign: MacroChildNode = {
+			type: 'mcfunction:macro_child/sign',
+			range: core.Range.create(start, start + 1),
+			value: '$',
+		}
+		children.push(sign)
+		src.skip()
+
+		// Handle the rest of the line
+
+		let beginning = src.cursor
+		let nextChunk = src.readUntil('$', '\r', '\n')
+		do { // Do-while because we always want it to work the first time
+			let wasMacro = false
+			const child: MacroChildNode = parseMacroChild(
+				src,
+				ctx,
+				nextChunk,
+				beginning,
+			)
+			if (child.type === 'mcfunction:macro_child/macro') {
+				wasMacro = true
+			}
+			// Handle a non-macro $ in the middle of the line
+			let subMacro: MacroChildNode | undefined
+			if (
+				child.type === 'mcfunction:macro_child/other' &&
+				child.value?.indexOf('$') !== -1
+			) {
+				const parts: String[] = child.value?.split('$') ?? []
+				let other = ''
+				for (const part of parts) {
+					if (part.charAt(0) == '(') { // We've found our macro key!
+						subMacro = parseMacroChild(
+							src,
+							ctx,
+							'$' + part,
+							beginning + other.length - 1,
+						)
+						child.range.end -= part.length + 1
+					} else { // Keep it in the normal string
+						other += part + '$'
+					}
+				}
+				wasMacro = true
+				child.value = other.substring(0, other.length - 1) // Remove the last '$'
+			}
+			children.push(child)
+			if (subMacro) { // Sub-macro added here because colorizer
+				children.push(subMacro)
+			}
+			// Prepare for the next block
+			beginning = src.cursor
+			if (wasMacro) {
+				nextChunk = src.readUntil('\r', '\n', '$')
+			} else {
+				nextChunk = src.readUntil('\r', '\n', ')')
+			}
+		} while (nextChunk.length > 0)
+
+		// A line with no macros is invalid
+		let hasMacro = false
+		for (const child of children) {
+			if (child.type === 'mcfunction:macro_child/macro') {
+				hasMacro = true
+				break
+			}
+		}
+		if (!hasMacro) {
+			ctx.err.report(
+				localize('expected', localize('macro')),
+				core.Range.create(start, src.cursor),
+			)
+		}
+
+		// Return the result
+		const ans: MacroNode = {
+			type: 'mcfunction:macro',
+			range: core.Range.create(start, src.cursor),
+			children: children,
+		}
+		return ans
+	}
+}
+
+/**
+ * Takes in a chunk of text read from the source and returns a MacroChildNode of type macro or other.
+ */
+function parseMacroChild(
+	src: core.Source,
+	ctx: core.ParserContext,
+	chunk: String,
+	beginning: number,
+): MacroChildNode {
+	if (chunk.substring(0, 2) === '$(') { // This is a macro key
```

Comment:
This function essentially does two completely unrelated things: validating a macro key (first if branch) and returning a node for the surrounding part of a macro line.

Reply:
Correct

Reply:
something you can do to make a complicated function more readable is to split the two underlying things its doing into 2 separate functions. (or more if you're feeling daring 😚)